src/compiler/binder.ts

```diff
@@ -278,6 +278,16 @@ namespace ts {
                     Debug.assert(isWellKnownSymbolSyntactically(nameExpression));
                     return getPropertyNameForKnownSymbolName(idText((<PropertyAccessExpression>nameExpression).name));
                 }
+                if (isPrivateName(name)) {
+                    // containingClass exists because private names only allowed inside classes
+                    const containingClass = getContainingClass(name.parent);
+                    if (!containingClass) {

```

Comment:
Currently, this errors in the parser, but we can change that:
Discussion:
https://github.com/microsoft/TypeScript/pull/30829/files/61b95d03685279d2188eaf117500fa2b2083d7e6#r299706018

Reply:
Added a Debug.fail. I agree it's best not to silently do the wrong thing if we get into this weird state.

Reply:
Whoops. Removing the Debug.fail breaks a test. Turns out the behavior without `Debug.fail` is closer to what I think we want:
In the case where this code path gets hit, the parser has already queued a DiagnosticMessage:

`error TS18016: Private names are not allowed outside class bodies.`

The error message from the parser seems better to me than a Debug.fail, since we get localized error messages and the compiler can keep going.